Adapt BlobMixin and all subclasses to new blobmeta API

millerdev opened the pull request:
Review commits individually 🐠 

@snopoke @mkangia

millerdev committed:
Adapt blob migrations to new blob metadata API

millerdev committed:
Add warning to blob storage report command

millerdev committed:
Adapt check_blob_logs management command to new blob metadata API

millerdev committed:
Adapt BlobMixin to new blob metadata API

millerdev committed:
Add blob db type code to InvoicePdf

millerdev committed:
Add blob db type code to ApplicationBase

millerdev committed:
Add blob db type code to CommCareBuild

millerdev committed:
Add blob db type code to Domain

millerdev committed:
Add blob db type code to ExportInstance

millerdev committed:
Add blob db type code to CommCareMultimedia

millerdev committed:
Add blob db type code to CommCareCase

millerdev committed:
Add blob db type code to SavedBasicExport

millerdev committed:
Add blob db type code to XFormInstance

millerdev committed:
Comment out old/obsolete migrations

millerdev committed:
Adapt tzmigration to new blob metadata API

millerdev committed:
Adapt ESAccessors tests to new blob metadata API

millerdev committed:
Adapt BlobHelper usages to blobmeta API

mkangia commented on corehq/blobs/mixin.py:
```diff
@@ -57,32 +76,55 @@ class Meta(object):
+            if normalized:
+                data = data.copy()
+                data["external_blobs"] = blobs
+        return super(BlobMixin, cls).wrap(data)
```
Looks like we are to consider just the ones that don't have `doc_type` as `BlobMetaPointer` and normalize them
can we re-write this as the following or a better version
```
if data.get("external_blobs"):
    if any(value["doc_type"] != "BlobMetaPointer" for key,value in data.get("external_blobs").items()):
        data = data.copy()
        doc_id = safe_id(data["_id"])
        dbname = _get_couchdb_name(cls)
        for key, value in data["external_blobs"].items():
            if value["doc_type"] != "BlobMetaPointer":
                data["external_blobs"][key] = normalize(dbname, doc_id, value)
return super(BlobMixin, cls).wrap(data)
```
to avoid iterating each time.

Also is this like a migration process? keep updating as an when accessed?

mkangia commented on corehq/blobs/mixin.py:
```diff
@@ -31,23 +32,41 @@
 
-class BlobMeta(DocumentSchema):
-    id = StringProperty()
+class BlobMetaPointer(DocumentSchema):
```
Have you considered `BlobMetaReference` as the name?

millerdev replied:
After thinking this over more, the term "pointer" could conjure ideas about the concept of pointers in C (the programming language), which isn't quite analogous here. Renamed to `BlobMetaRef` in c50c4eed51f69366d0a71d7bb3b76a6f46151a6b

mkangia commented on corehq/blobs/mixin.py:
```diff
@@ -96,30 +138,44 @@ def put_attachment(self, content, name=None, content_type=None, content_length=N
             raise InvalidAttachment("cannot save attachment without name")
+        if self._id is None:
+            raise ResourceNotFound("cannot put attachment on unidentified document")
+        if hasattr(self, "domain"):
```
Why is this check needed?

millerdev replied:
Some objects that use `put_attachment` do not have a `domain` attribute. Example: `InvoicePdf`

https://github.com/dimagi/commcare-hq/blob/25d83a6250abcbde256ef275ef2145f237929873/corehq/apps/accounting/models.py#L2883

The blob db now requires a domain when saving blobs. There are two ways the domain may be provided here:

- `domain` argument passed to this method (must match `self.domain` if the attribute exists).
- no `domain` passed -> use `self.domain` value (if present), otherwise raise an error.

mkangia commented on corehq/blobs/mixin.py:
```diff
@@ -261,34 +315,41 @@ class BlobHelper(object):
+            return blobs
+        doc_id = doc["_id"]
+        return {n: BlobMetaPointer._normalize_json(dbname, doc_id, m)
+                for n, m in six.iteritems(blobs)}
```
Can this be
```
if "external_blobs" not in doc:
    return default
doc_id = doc["_id"]
return {n: BlobMetaPointer._normalize_json(dbname, doc_id, m) for n, m in six.iteritems(blobs)}
```

millerdev replied:
That wouldn't work because `six.iteritems(blobs)` would cause a NameError

snopoke replied:
I agree with @mkangia that his version requires less brain power to understand. Also I think you're missing the point, you'd do `six.iteritems(doc['external_blobs'])`

millerdev replied:
This method (`get_external_blobs`) got refactored and doesn't exist anymore.

mkangia replied:
thats even better :)

mkangia commented on corehq/blobs/util.py:
```diff
@@ -84,6 +84,16 @@ def __get__(self, obj, owner):
         return self.func.__get__(obj, owner)
 
 
+class classproperty(object):
```
interesting.

mkangia commented on corehq/apps/accounting/models.py:
```diff
@@ -2874,12 +2876,13 @@ def generate_pdf(self, invoice):
 
         template.get_pdf()
         filename = self.get_filename(invoice)
+        blob_domain = domain or "<unknown>"
```
nit: how about extracting `<unknown>` as a const like, `UNKNOWN_DOMAIN_PLACEHOLDER` or something similar instead of keeping a string here which can possibly be used again or easy to change later at a single spot if needed?

mkangia replied:
And is this expected? to have an invoice without domain?

mkangia commented on corehq/apps/domain/models.py:
```diff
@@ -1016,6 +1019,9 @@ def get_custom_logo(self):
         )
 
+    def put_attachment(self, *args, **kw):
+        return super(Domain, self).put_attachment(domain=self.name, *args, **kw)
```
is this correct? The expected arguments looks to be different for `BlobMixin`, there is content as the first argument?

mkangia commented on corehq/apps/hqmedia/models.py:
```diff
@@ -139,7 +141,12 @@ def attach_data(self, data, original_filename=None, username=None, attachment_id
+                data,
+                attachment_id,
+                content_type=self.get_mime_type(data, filename=original_filename),
+                domain=SHARED_DOMAIN,
```
Just confirming that this is shared because the file_hash will be the same in case the same one is used for different domains.

Just one follow up question. In case we need to find blobs for a domain this wont be included then? like when we filter them in SQL, i guess that's how we would filter after this change?

mkangia commented on corehq/ex-submodules/couchexport/models.py:
```diff
@@ -969,7 +970,11 @@ def get_attachment_name(self):
+        # future, so I didn't think it was worth it to try to pass the domain
+        # in here.
+        domain = "<unknown>"
+        self.put_attachment(payload, self.get_attachment_name(), domain=domain)
```
what would be the difference in passing the `domain` though?

millerdev replied:
It's required by `blobdb.put(...)`, but `SavedBasicExport` doesn't have a `domain` attribute.

mkangia replied:
Interesting. I never thought there would be exports not bound to a domain.

mkangia reviewed (commented):
Could not review the complete PR yet, sharing comments/queries/suggestions on how far i got.

millerdev committed:
Temporarily fix Couch -> SQL migration script

This will be fixed correctly when SQL form attachments are converted to use the new blobmeta API.

millerdev committed:
Use UNKNOWN_DOMAIN constant

millerdev committed:
Fix Couch -> SQL migration again

Not sure how the null value is slipping through.

millerdev committed:
Fix ExportInstance tests

millerdev commented:
✅

snopoke commented on corehq/blobs/mixin.py:
```diff
@@ -57,32 +76,55 @@ class Meta(object):
+                    blobs[key] = normalize(dbname, doc_id, value)
+                    normalized = True
+            if normalized:
+                data = data.copy()
```
why is the `copy` necessary?

snopoke replied:
Yea I know that's generally a good idea. In this case the data is coming from an HTTP request so modifying it shouldn't really be an issue. The reason I brought it up is that it's going to double the memory requirement for any doc loaded (which in the case of largs apps could be significant).

snopoke replied:
The other thing is that JsonObject mutates the dictionary under the hood anyway.

millerdev replied:
It's only a shallow copy, so it will not result in that much extra memory use. Also, just because `JsonObject` does a bad thing doesn't make me feel better about doing it in other places—the hard-to-find bugs get harder to find when it's done in multiple places.

snopoke commented on corehq/ex-submodules/couchforms/models.py:
```diff
@@ -137,6 +137,7 @@ class XFormInstance(DeferredBlobMixin, SafeSaveDocument, UnicodeMixIn,
     date_header = DefaultProperty()
     build_id = StringProperty()
     export_tag = DefaultProperty(name='#export_tag')
+    _blobdb_type_code = CODES.form_xml
```
forms an have `form_xml` and `form_attachment` right? How is that handled?

millerdev replied:
Good catch. eb85c0992c54ece89c70879ab7c83657e8d91e0e

snopoke commented on corehq/ex-submodules/casexml/apps/case/models.py:
```diff
@@ -178,8 +178,9 @@ class CommCareCase(DeferredBlobMixin, SafeSaveDocument, IndexHoldingMixIn,
-    
+
     server_modified_on = DateTimeProperty()
+    _blobdb_type_code = CODES._default
```
If this is only used for cases might as well call it `_case`? Or are there other places you plan to use it?

snopoke reviewed (approved):
Couple of questions / comments but nothing blocking.

millerdev committed:
Use correct type code for couch xform attachments

millerdev committed:
BlobMetaPointer -> BlobMetaRef

mkangia commented on corehq/apps/couch_sql_migration/couchsqlmigration.py:
```diff
@@ -486,7 +486,7 @@ def _migrate_form_attachments(sql_form, couch_form):
             blob_id=blob.key,
             blob_bucket="",  # temporary/special value -> new blob db API
-            md5=blob.info.md5_hash,
+            md5=blob.info.md5_hash or "unknown-md5",
```
Just checking, why is this stub/constant value okay here to pass?

millerdev replied:
The new blob db API does not use or maintain the MD5 hash, so in places where we don't have one and the value cannot be `None`, I used a value with obvious meaning if someone happens to read it.

millerdev committed:
Merge branch 'master' into dm/blobmeta-mixin
